### PR TITLE
Support of Horovod and TF 1.12 for TensorFlow Script Mode. TFS 1.12 support

### PR DESCRIPTION
**Pull request by mvsusp**
*Description of changes:*
* feature: support for Tensorflow 1.12
* feature: support for Tensorflow Serving 1.12
* feature: support for Horovod

## Merge Checklist

_Put an `x` in the boxes that apply. You can also fill these out after creating the PR. If you're unsure about any of them, don't hesitate to ask. We're here to help! This is simply a reminder of what we are going to look for before merging your pull request._

- [x] I have read the [CONTRIBUTING](https://github.com/aws/sagemaker-python-sdk/blob/master/CONTRIBUTING.md) doc
- [x] I have added tests that prove my fix is effective or that my feature works (if appropriate)
- [x] I have updated the [changelog](https://github.com/aws/sagemaker-python-sdk/blob/master/CHANGELOG.rst) with a description of my changes (if appropriate)
- [x] I have updated any necessary [documentation](https://github.com/aws/sagemaker-python-sdk/blob/master/README.rst) (if appropriate)

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: .gitmodules [hunk 1]
```diff
@@ -0,0 +1,4 @@
+[submodule "tests/data/horovod/benchmarks"]
```
Comment (nadiaya) [inline .gitmodules]: I don't think we should be adding submodule

File: src/sagemaker/tensorflow/defaults.py [hunk 2]
```diff
@@ -12,4 +12,4 @@
 # language governing permissions and limitations under the License.
 from __future__ import absolute_import
 
-TF_VERSION = '1.11'
```
Comment (nadiaya) [inline src/sagemaker/tensorflow/defaults.py]: Don't change default version.

Add LATEST_VERSION to the estimator instead, e.g.: https://github.com/aws/sagemaker-python-sdk/blob/master/src/sagemaker/mxnet/estimator.py#L33

File: tests/data/horovod/launcher.sh [hunk 3]
```diff
@@ -0,0 +1,3 @@
+#!/usr/bin/env bash
+
+python benchmarks/scripts/tf_cnn_benchmarks/tf_cnn_benchmarks.py --num_batches=500 --model vgg16 --variable_update horovod --horovod_device gpu --use_fp16 --summary_verbosity 1 --save_summaries_steps 10         --train_dir /opt/ml/model --eval_dir /opt/ml/model --batch_size 32
```
Comment (nadiaya) [inline tests/data/horovod/launcher.sh]: new line

File: src/sagemaker/tensorflow/estimator.py [hunk 4]
```diff
@@ -200,14 +200,21 @@ def __init__(self, training_steps=None, evaluation_steps=None, checkpoint_path=N
             script_mode (bool): If set to True will the estimator will use the Script Mode containers (default: False).
                 This will be ignored if py_version is set to 'py3'.
             distributions (dict): A dictionary with information on how to run distributed training
-                (default: None). Currently we only support distributed training with parameter servers. To enable it
-                use the following setup:
+                (default: None). Currently we support distributed training with parameter servers and MPI. To enable
+                parameter servers use the following setup:
```
Comment (nadiaya) [inline src/sagemaker/tensorflow/estimator.py]: Follow up:
we should have a test that checks parmater server + mpi works.

File: src/sagemaker/tensorflow/estimator.py [hunk 5]
```diff
@@ -419,13 +426,24 @@ def hyperparameters(self):
...
+                additional_hyperparameters[self.LAUNCH_PS_ENV_NAME] = ps_enabled
+
+            if 'mpi' in self.distributions:
+                mpi_dict = self.distributions['mpi']
+                mpi_enabled = mpi_dict.get('enabled', False)
+                additional_hyperparameters[self.LAUNCH_MPI_ENV_NAME] = mpi_enabled
+                additional_hyperparameters[self.MPI_NUM_PROCESSES_PER_HOST] = mpi_dict.get('processes_per_host', 1)
+                additional_hyperparameters[self.MPI_CUSTOM_MPI_OPTIONS] = mpi_dict.get('custom_mpi_options', '')
+
+            self.model_dir = self.model_dir or self._default_s3_path('model', mpi=mpi_enabled)
```
Comment (nadiaya) [inline src/sagemaker/tensorflow/estimator.py]: isn't it s3 checkpointing dir?
why is it equals to /opt/ml/model for mpi? shouldn't it be the same for mpi and parameter server distributed cases?

File: tests/integ/test_horovod.py [hunk 6]
```diff
@@ -0,0 +1,68 @@
...
+from sagemaker.tensorflow import TensorFlow
+from six.moves.urllib.parse import urlparse
+import tests.integ as integ
+from tests.integ import timeout, vpc_test_utils
+
+horovod_dir = os.path.join(os.path.dirname(__file__), '..', 'data', 'horovod')
+
+
+@pytest.mark.parametrize('instance_type', ['ml.p3.2xlarge', 'ml.p2.xlarge'])
+def test_horovod_tf_benchmarks(sagemaker_session, instance_type):
```
Comment (nadiaya) [inline tests/integ/test_horovod.py]: discussed online - i believe this test should live in tf-container, too heavy for sdk testing purposes.

File: tests/data/horovod/launcher.sh (hunk 3 above)
Comment (uditbhatia) [reply tests/data/horovod/launcher.sh]: done

File: src/sagemaker/tensorflow/defaults.py (hunk 2 above)
Comment (uditbhatia) [reply src/sagemaker/tensorflow/defaults.py]: For Mxnet also the default version in defaults.py and LATEST_VERSION in estimator.py is explicityly defined. 
MxNet estimatopr.py => https://github.com/aws/sagemaker-python-sdk/blob/master/src/sagemaker/mxnet/estimator.py#L33
MxNet defaults.py -> https://github.com/aws/sagemaker-python-sdk/blob/master/src/sagemaker/mxnet/defaults.py

File: CHANGELOG.rst [hunk 7]
```diff
@@ -6,12 +6,21 @@ CHANGELOG
 ======
 
 * enhancement: Workflow: Specify tasks from which training/tuning operator to transform/deploy in related operators
+* feature: support for Tensorflow 1.12
+* feature: support for Tensorflow Serving 1.12
```
Comment (mvsusp) [inline CHANGELOG.rst]: Remove repeated features from changelog

File: src/sagemaker/tensorflow/README.rst [hunk 8]
```diff
@@ -207,6 +207,64 @@ After attaching, the estimator can be deployed as usual.
...
+
+The container sets up the MPI environment and executes the ``mpirun`` command enabling you to run any Horovod
+training script with Script Mode.
+
+Training with ``MPI`` is configured by specifying following fields in ``distributions``:
+
+- ``enabled (bool)``: If set to ``True``, the MPI setup is performed and ``mpirun`` command is executed.
+- ``processes_per_host (int)``: Number of processes MPI should launch on each host. Note, this should not be
+  greater than the available slots on the selected instance type.
+- ``custom_mpi_options (str)``: Additional command line arguments to pass to ``mpirun``.
```
Comment (mvsusp) [inline src/sagemaker/tensorflow/README.rst]: Please detail more this setting, or link to the notebook example.

File: src/sagemaker/tensorflow/README.rst [hunk 9]
```diff
@@ -207,6 +207,64 @@ After attaching, the estimator can be deployed as usual.
...
+
+In the below example we create an estimator to launch Horovod distributed training with 2 processes on one host:
+
+.. code:: python
+
+    from sagemaker.tensorflow import TensorFlow
+
+    tf_estimator = TensorFlow(entry_point='tf-train.py', role='SageMakerRole',
+                              train_instance_count=1, train_instance_type='ml.p2.xlarge',
+                              framework_version='1.11', py_version='py3',
```
Comment (mvsusp) [inline src/sagemaker/tensorflow/README.rst]: Horovod is only available on 1.12 and ahead.

File: src/sagemaker/tensorflow/estimator.py (hunk 5 above)
Comment (mvsusp) [reply src/sagemaker/tensorflow/estimator.py]: Horovod does not need S3 for distributed training. /opt/ml/model is the correct path here. I will resolve this conversation.

File: tests/data/mxnet_mnist/MANIFEST.in [hunk 10]
```diff
@@ -0,0 +1,5 @@
+
```
Comment (mvsusp) [inline tests/data/mxnet_mnist/MANIFEST.in]: It seems that some files on this PR, including this one, should be there. Maybe is it a result of a bad merge?

Please remove these files.

File: tests/integ/test_tf_script_mode.py [hunk 11]
```diff
@@ -82,3 +85,18 @@ def _assert_s3_files_exist(s3_url, files):
         found = [x['Key'] for x in contents if x['Key'].endswith(f)]
         if not found:
             raise ValueError('File {} is not found under {}'.format(f, s3_url))
+
+
+def _assert_s3_files_exist_in_tar(s3_url, files):
```
Comment (mvsusp) [inline tests/integ/test_tf_script_mode.py]: This function is not being used anywhere.

File: CHANGELOG.rst (hunk 7 above)
Comment (uditbhatia) [reply CHANGELOG.rst]: done

File: tests/data/mxnet_mnist/MANIFEST.in (hunk 10 above)
Comment (uditbhatia) [reply tests/data/mxnet_mnist/MANIFEST.in]: Removed

File: tests/integ/test_tf_script_mode.py (hunk 11 above)
Comment (uditbhatia) [reply tests/integ/test_tf_script_mode.py]: Remoived

File: CHANGELOG.rst [hunk 12]
```diff
@@ -7,7 +7,6 @@ CHANGELOG
 
 * enhancement: Workflow: Specify tasks from which training/tuning operator to transform/deploy in related operators
 * feature: support for Tensorflow 1.12
```
Comment (mvsusp) [inline CHANGELOG.rst]: Remove support for TF 1.12 as well (repeated)

File: CHANGELOG.rst [hunk 13]
```diff
@@ -6,12 +6,21 @@ CHANGELOG
...
 
+
 1.17.0
 ======
 
+
 * bug-fix: Workflow: Revert appending Airflow retry id to default job name
+* feature: support for Tensorflow 1.12
+* feature: support for Tensorflow Serving 1.12
+* feature: support for Horovod
```
Comment (mvsusp) [inline CHANGELOG.rst]: Repeated.

File: src/sagemaker/tensorflow/defaults.py (hunk 2 above)
Comment (mvsusp) [reply src/sagemaker/tensorflow/defaults.py]: Don't change the version @uditbhatia . Version is a required field starting by TF 1.12.

Comment (uditbhatia) [reply src/sagemaker/tensorflow/defaults.py]: I didnt changed any field

File: CHANGELOG.rst (hunk 12 above)
Comment (uditbhatia) [reply CHANGELOG.rst]: done

File: CHANGELOG.rst (hunk 13 above)
Comment (uditbhatia) [reply CHANGELOG.rst]: removed

**Pull request merged**